PEP 734: Multiple Interpreters in the Stdlib (#3523)
python/peps: peps/pep-0734.rst

purposefully keeping the initial API as lean and minimal as possible.
The objective is to provide a well-considered foundation on which we may
add further (more advanced) functionality later.

That said, the proposed design incorporates lessons learned from
existing use of subinterpreters by the community, from existing stdlib
modules, and from other programming languages.  It also factors in
experience from using subinterpreters in the CPython test suite and
using them in `concurrency benchmarks`_.

.. _concurrency benchmarks:
   https://github.com/ericsnowcurrently/concurrency-benchmarks

The module will include a basic mechanism for communicating between
interpreters.  Without one, multiple interpreters are a much less
useful feature.

Specification
=============

The module will:

* expose the existing multiple interpreter support
* introduce a basic mechanism for communicating between interpreters

The module will wrap a new low-level ``_interpreters`` module
(in the same way as the ``threading`` module).  However, that low-level
API is not intended for public use and thus not part of this proposal.

We also expect that an ``InterpreterPoolExecutor`` will be added to the

> ``concurrent.futures`` module, but that is outside the scope of this PEP.

Author: I don't mind pulling it in.

> ``concurrent.futures`` module, but that is outside the scope of this PEP.

Reviewer: Great. In the previous PEP version, this didn't follow the Executor model very closely -- instead of a function and optional arguments you were passing a string (like `Interpreter.exec()`) IIUC `ProcessPoolExecutor` has solved this issue, so hopefully `InterpreterPoolExecutor` can solve it similarly. (I'm guessing using `pickle`? Though I've never looked.)

> ``concurrent.futures`` module, but that is outside the scope of this PEP.

Author: Yeah, it would be pickle.

| ``.is_running() -> bool``        | Is the interpreter currently executing code?   |
+----------------------------------+------------------------------------------------+
| ``.set_main_attrs(**kwargs)``    | Bind objects in ``__main__``.                  |
+----------------------------------+------------------------------------------------+
| ``.exec(code, /)``               | | Run the given source code in the interpreter |
|                                  | | (in the current thread).                     |
+----------------------------------+------------------------------------------------+

Additional details:

* Every ``Interpreter`` instance adds a reference to the corresponding
  interpreter, which is released when the ``Interpreter`` instance is
  deleted.  When there are no more such references to an interpreter,
  it gets finalized.  Thus no interpreters created through
  ``interpreters.create()`` will leak.

* ``Interpreter.is_running()`` refers only to if there is a thread
  running a script (code) in the interpreter's ``__main__`` module.
  That basically means whether or not ``Interpreter.exec()`` is running

>   in some thread.  Code running in sub-threads is ignored.

Author: The main practical value I see is being able to query before calling `interp.exec()`, to avoid getting an exception if it's already running.  Otherwise it's similar to `Thread.is_alive()` and would have similar utility.  I also imagine it could be helpful for people poking around and exploring how interpreters work.

  be useful for running some code to initialize an interpreter
  (e.g. with imports) before later performing some repeated task.

Comparison with builtins.exec()
-------------------------------

``Interpreter.exec()`` is essentially the same as the builtin
``exec()``, except it targets a different interpreter, using that
interpreter's isolated state.

The builtin ``exec()`` runs in the current OS thread and pauses
whatever was running there, which resumes when ``exec()`` finishes.
No other threads are affected.  (To avoid pausing the current thread,
run ``exec()`` in a ``threading.Thread``.)

``Interpreter.exec()`` works the same way.

The builtin ``exec()`` executes against a namespace, by default the
``__dict__`` of the current module (i.e. ``globals()``).
It uses that namespace as-is and does not clear it before or after.

``Interpreter.exec()`` works the same way.

...where the "current" module is always the ``__main__`` module
of the target interpreter.  This is the same as how it works
for code run from the Python command-line (e.g. ``-m``) or the REPL.

The builtin ``exec()`` discards any object returned from the

> executed code.

Author: I agree, there is no way to return an object from `exec()`.  Any return value is discarded.  Normally what one passes to `exec()` wouldn't have a `return` statement.  However, it actually can:

```py
def func():
    global spam
    spam = True
    return 'a returned value'

ns = {}
res = exec(func.__code__, ns)
# Nothing is returned.  The returned string was discarded.
assert res is None, res
assert ns['spam'] is True, ns
```

> executed code.

Reviewer: Oh, I had forgotten about that. It might help if you called out this edge case. :-)

``Interpreter.exec()`` works the same way.

The builtin ``exec()`` propagates any uncaught exception from the code
it ran.  The exception is raised from the ``exec()`` call in the
thread that originally called ``exec()``.

``Interpreter.exec()`` works the same way.

...with one slight difference.  Rather than propagate the uncaught
exception directly, we raise an ``interpreters.RunFailedError``

> with a snapshot of the uncaught exception (including its traceback)

Author: I had planned on creating a lightweight custom class for the snapshot, but @iritkatriel pointed out to me that I could just use `traceback.TracebackException`.  That works pretty well.

> with a snapshot of the uncaught exception (including its traceback)

Author: This became an option recently when @pablogsal cleaned up the default `sys.excepthook` to use `traceback.TracebackException`: gh-110702.

> as the ``__cause__``.  Directly raising (a proxy of) the exception

Author: If we make `traceback.TracebackException` inherit from `Exception` then this is a non-issue.  (The same goes if we use some other type for the snapshot.)

is problematic since it's harder to distinguish between an error
in the ``Interpreter.exec()`` call and an uncaught exception
from the subinterpreter.

API: Communicating Between Interpreters
---------------------------------------

The module introduces a basic communication mechanism called "channels".
They are based on `CSP`_, as is Go's concurrency model (loosely).

> Channels are like pipes: FIFO queues with distinct send/receive ends.

Author: I'm not quite sure what you mean by "1:1 correspondence".  Each sent object is added to the back of the channel's internal queue and each `recv()` pops the next one off the front of that queue.  Every sent object will be received.

> Channels are like pipes: FIFO queues with distinct send/receive ends.

Reviewer: Okay, sorry, yes, that's what I meant by 1:1 correspondence. For pipes this isn't the case, unless you implement some higher-level protocol: when using pipes, one end can write two lines using two `write()` calls, and the other end *may* receive them all in a single `read()` call (and data written using a single `write()` call may be spread across multiple `read()` calls).

Maybe it's better not to mention pipes and just state that channels are queues?

(Which reminds me -- channels are thread-safe, right?)

> Channels are like pipes: FIFO queues with distinct send/receive ends.

Author: Yeah, channels are completely thread-safe.

They are designed to work safely between isolated interpreters.

.. _CSP:
   https://en.wikipedia.org/wiki/Communicating_sequential_processes

For now, only objects that are specifically supported for passing
between interpreters may be sent through a channel.
See `Shareable Objects`_.

The module's top-level API for this new mechanism:

+----------------------------------------------------+-----------------------+
| signature                                          | description           |
+====================================================+=======================+

> | ``create_channel() -> (RecvChannel, SendChannel)`` | Create a new channel. |

Author: `RecvChannel` is the receiving end and `SendChannel` is the sending end.  Perhaps it would help if they were something like `RecvChannelEnd` and `SendChannelEnd`?

> | ``create_channel() -> (RecvChannel, SendChannel)`` | Create a new channel. |

Reviewer: Hm, those don't delight me, but I can't think of something better either. I wonder if something using sender/receiver or reader/writer could be used?

The key insight that I eventually reached was that this returns two ends of a queue, and the API is a bit like `r, w = os.pipe()`. Including the part where you need to "give" the other party the other end of a channel you created. Which you can do using either `set_main_attr()` or by sending it over an already established channel. And that's why channels are "shareable" objects (rather uniquely, since all the other shareable types are basic data types).

> | ``create_channel() -> (RecvChannel, SendChannel)`` | Create a new channel. |

Author: Honestly, I'm probably going to drop the whole "distinct ends" thing, have one class, and just call it `Queue`, in the spirit of the minimal API I'm aiming for.  For that matter, I suppose we could put it in the `queue` module (something like `CrossInterpreterQueue`), but it probably still belongs in `interpreters`.  We could add directional channels later, if worth doing.

My only reservation is that single-directional channel ends often makes code easier to reason about and harder to do the wrong thing.  A bi-directional queue object has the opposite effect.  At least, that was my experience with Go, which has both directional and bi-directional channels.  A bi-directional channel was never necessary but it was easy to use one without thinking (at the expense of extra cognitive overhead later), since that's what you get when you create a channel (e.g. `var foo = make(chan int)`).  Having only directional channels from the start eliminates the problem, which is part of why I've been proposing `RecvChannel` and `SendChannel`.  In some ways it's better to not introduce the less desirable thing in the first place, rather than try to discourage its use (or remove it) later.

However, given your reaction, I'm inclined to think it isn't worth having the distinct ends.  Furthermore, I don't recall anyone bringing up concerns with Python's `queue.Queue` like I've experienced with Go's bi-directional channel.  So it may very well be a Go-only issue and one I needn't worry about with the `interpreters` module.

> | ``create_channel() -> (RecvChannel, SendChannel)`` | Create a new channel. |

Reviewer: A single Queue class sounds good, just don't put it in queue.py (it's not really related except at a high level. For example, asyncio has its own queue submodule.

I wouldn't worry too much about not having distinct read/write ends -- we live with that for file I/O too, for example. I presume that using a Queue, an interpreter could possibly send to itself, which seems fine as an edge case.

Is there something about queue capacity?

> | ``create_channel() -> (RecvChannel, SendChannel)`` | Create a new channel. |

Author: > Is there something about queue capacity?

I didn't see a need to worry about anything but unbounded for now.

> | ``create_channel() -> (RecvChannel, SendChannel)`` | Create a new channel. |

Reviewer: Okay, but then why do we have both `send()` and `send_nowait()`? If the queue is unbounded there's never a need to block in `send()`, is there? How is the difference implemented?

(I have no beef with `recv_nowait()` -- if the queue is empty, one blocks, the other doesn't. But in queue.py, `put_nowait()` exists to avoid blocking when the queue is *full*.)

> | ``create_channel() -> (RecvChannel, SendChannel)`` | Create a new channel. |

Author: In that context you can think of the queue as having a max size of 0 then.

FWIW, I've found it useful to know when the object is actually popped off the queue, as a synchronization point between two interpreters.

> | ``create_channel() -> (RecvChannel, SendChannel)`` | Create a new channel. |

Reviewer: Oh wait, the updated description makes it clear that queues have max capacity zero. That's surprising given that previously you described it as similar to `os.pipe()`, but it is a consistent model, so I withdraw my objections.

Now I vaguely wonder how we'd implement a queue *with* buffering on top of this, but that's a straightforward programming problem.

|                              | | to the receiving end of the channel and wait.                     |
+------------------------------+---------------------------------------------------------------------+
| ``.send_nowait(obj)``        | Like ``send()``, but return False if not received.                  |
+------------------------------+---------------------------------------------------------------------+

Shareable Objects
-----------------

Both ``Interpreter.set_main_attrs()`` and channels work only with
"shareable" objects.

A "shareable" object is one which may be passed from one interpreter
to another.  The object is not necessarily actually shared by the
interpreters.  However, the object in the one interpreter is guaranteed
to exactly match the corresponding object in the other interpreter.

For some types, the actual object is shared.  For some, the object's
underlying data is actually shared but each interpreter has a distinct
object wrapping that data.  For all other shareable types, a strict copy
or proxy is made such that the corresponding objects continue to match.

For now, shareable objects must be specifically supported internally

> by the Python runtime.

Author: I expect sharing mutable data directly (without a proxy) will be uncommon  Regardless, shared mutable data ends up a matter of thread-safety, and that would have to be the responsibility of the user.  For `memoryview` (and buffers generally), I expect it to be manageable.

I'll make sure the PEP is clear about this.

> by the Python runtime.

Reviewer: Do you have locking primitives so users *can* serialize access to shared data if they want to go that way?

> by the Python runtime.

Author: Channels are basically the only locking primitive.  You pass a proverbial football around.

> by the Python runtime.

Reviewer: That's worth writing down.

> by the Python runtime.

Author: I'll add a short section about synchronization.

   This PEP is a reduction back to the essential information.  Much of
   that extra information is still valid and useful, just not in the
   immediate context of the specific proposal here.

Abstract
========

I propose that we add a new module, "interpreters", to the standard
library, to make the existing multiple-interpreters feature of CPython
more easily accessible to Python code.  This is particularly relevant
now that we have a per-interpreter GIL (:pep:`684`) and people are
more interested in using multiple interpreters.  Without a stdlib
module, users are limited to the `C-API`_, which restricts how much
they can try out and take advantage of multiple interpreters.

.. _C-API:
   https://docs.python.org/3/c-api/init.html#sub-interpreter-support

Rationale
=========

The ``interpreters`` module will provide a high-level interface to the
multiple interpreter functionality.  Since we have no experience with
how users will make use of multiple interpreters in Python code, we are
purposefully keeping the initial API as lean and minimal as possible.
The objective is to provide a well-considered foundation on which we may

> add more-advanced functionality later.

Reviewer 2: ```suggestion
add more advanced functionality later.
```

Author: The hyphen was intentional, but your comment demonstrates that it may be awkward. 😄  I'll reword the phrase.

| signature                        | description                                    |
+==================================+================================================+
| ``class Interpreter``            | A single interpreter.                          |
+----------------------------------+------------------------------------------------+
| ``.id``                          | The interpreter's ID (read-only).              |
+----------------------------------+------------------------------------------------+
| ``.is_running() -> bool``        | Is the interpreter currently executing code?   |
+----------------------------------+------------------------------------------------+
| ``.set_main_attrs(**kwargs)``    | Bind objects in ``__main__``.                  |
+----------------------------------+------------------------------------------------+
| ``.exec(code, /)``               | | Run the given source code in the interpreter |
|                                  | | (in the current thread).                     |
+----------------------------------+------------------------------------------------+

Additional details:

> * Every ``Interpreter`` instance wraps an ``InterpreterID`` object.

Reviewer 2: Is this a public type? Is it the type of the `.id` attribute mentioned above?

Author: It is a public type and will be exposed as `interpreters.InterpreterID`.  However, I wasn't planning on including that in the PEP.  I'll reword the paragraph.

PEP: 734
Title: Multiple Interpreters in the Stdlib
Author: Eric Snow <ericsnowcurrently@gmail.com>
Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Created: 06-Nov-2023
Python-Version: 3.13
Replaces: 554

.. note::
   This PEP is essentially a continuation of :pep:`554`.  That document
   had grown a lot of ancillary information across 7 years of discussion.
   This PEP is a reduction back to the essential information.  Much of
   that extra information is still valid and useful, just not in the
   immediate context of the specific proposal here.

Abstract
========

I propose that we add a new module, "interpreters", to the standard
library, to make the existing multiple-interpreters feature of CPython
more easily accessible to Python code.  This is particularly relevant
now that we have a per-interpreter GIL (:pep:`684`) and people are
more interested in using multiple interpreters.  Without a stdlib

> module, users are limited to the `C-API`_, which restricts how much
> they can try out and take advantage of multiple interpreters.
> 
> .. _C-API:
>    https://docs.python.org/3/c-api/init.html#sub-interpreter-support

Reviewer 3: Using Intersphinx:
```suggestion
module, users are limited to the :ref:`C API <python:sub-interpreter-support>`, which restricts how much
they can try out and take advantage of multiple interpreters.
```

| ``class Interpreter``            | A single interpreter.                          |
+----------------------------------+------------------------------------------------+
| ``.id``                          | The interpreter's ID (read-only).              |
+----------------------------------+------------------------------------------------+
| ``.is_running() -> bool``        | Is the interpreter currently executing code?   |
+----------------------------------+------------------------------------------------+
| ``.set_main_attrs(**kwargs)``    | Bind objects in ``__main__``.                  |
+----------------------------------+------------------------------------------------+
| ``.exec(code, /)``               | | Run the given source code in the interpreter |
|                                  | | (in the current thread).                     |
+----------------------------------+------------------------------------------------+

Additional details:

* Every ``Interpreter`` instance wraps an ``InterpreterID`` object.
  When there are no more references to an interpreter's ID, it gets
  finalized.  Thus no interpreters created through
  ``interpreters.create()`` will leak.

|

> 

Author: No.  In my editor the bullets felt a bit cramped.  However, in the preview I see it shouldn't be a problem.

...with one slight difference: the namespace is implicit
(the ``__main__`` module's ``__dict__``).  This is the same as how
scripts run from the Python commandline or REPL work.

The builtin ``exec()`` discards any object returned from the
executed code.

``Interpreter.exec()`` works the same way.

The builtin ``exec()`` propagates any uncaught exception from the code
it ran.  The exception is raised from the ``exec()`` call in the
thread that originally called ``exec()``.

``Interpreter.exec()`` works the same way.

...with one slight difference.  Rather than propagate the uncaught
exception directly, we raise an ``interpreters.RunFailedError``
with a snapshot of the uncaught exception (including its traceback)
as the ``__cause__``.  Directly raising (a proxy of) the exception
is problematic since it's harder to distinguish between an error
in the ``Interpreter.exec()`` call and an uncaught exception
from the subinterpreter.

API: Communicating Between Interpreters
---------------------------------------

The module introduces a basic communication mechanism called "channels".
They are based on `CSP`_, as is ``Go``'s concurrency model (loosely).

> Channels are like pipes: FIFO queues with distinct send/recv ends.

Reviewer 3: ?
```suggestion
Channels are like pipes: FIFO queues with distinct send/receive ends.
```

Author: I was mirroring the type and method names.  However, that doesn't add much here.  I'll spell it out.

| ``create_channel() -> (RecvChannel, SendChannel)`` | Create a new channel. |
+----------------------------------------------------+-----------------------+

The objects for the two ends of a channel:

+------------------------------------------+-----------------------------------------------+
| signature                                | description                                   |
+==========================================+===============================================+
| ``class RecvChannel(id)``                | The receiving end of a channel.               |
+------------------------------------------+-----------------------------------------------+
| ``.id``                                  | The channel's unique ID.                      |
+------------------------------------------+-----------------------------------------------+
| ``.recv() -> object``                    | | Get the next object from the channel,       |
|                                          | | and wait if none have been sent.            |
+------------------------------------------+-----------------------------------------------+

> | ``.recv_nowait(default=None) -> object`` | | Like recv(), but return the default         |

Reviewer 3: ```suggestion
| ``.recv_nowait(default=None) -> object`` | | Like ``recv()``, but return the default     |
```

+------------------------------------------+-----------------------------------------------+

|

+------------------------------+---------------------------------------------------------------------+
| signature                    | description                                                         |
+==============================+=====================================================================+
| ``class SendChannel(id)``    | The sending end of a channel.                                       |
+------------------------------+---------------------------------------------------------------------+
| ``.id``                      | The channel's unique ID.                                            |
+------------------------------+---------------------------------------------------------------------+
| ``.send(obj)``               | | Send the `shareable object <Shareable Objects_>`_ (i.e. its data) |
|                              | | to the receiving end of the channel and wait.                     |
+------------------------------+---------------------------------------------------------------------+

> | ``.send_nowait(obj)``        | Like send(), but return False if not received.                      |

Reviewer 3: ```suggestion
| ``.send_nowait(obj)``        | Like ``send()``, but return ``False`` if not received.              |
```

interpreters.  However, the object in the one interpreter is guaranteed
to exactly match the corresponding object in the other interpreter.

For some types, the actual object is shared.  For some, the object's
underlying data is actually shared but each interpreter has a distinct
object wrapping that data.  For all other shareable types, a strict copy
or proxy is made such that the corresponding objects continue to match.

For now, shareable objects must be specifically supported internally
by the Python runtime.

Here's the initial list of supported objects:

* ``str``
* ``bytes``
* ``int``
* ``float``
* ``bool`` (``True``/``False``)
* ``None``
* ``tuple`` (only with shareable items)
* channels (``SendChannel``/``RecvChannel``)
* ``memoryview``

Again, for some types the actual object is shared, whereas for others
only the underlying data (or even a copy or proxy) is shared.
Eventually mutable objects may also be shareable.

Regardless, the guarantee of "shareable" objects is that corresponding
objects in different interpreters will always strictly match each other.

Examples
--------

Using interpreters as workers, with channels to communicate:

::

>    tasks_recv, tasks = interpreters.create_channel()

Reviewer 3: ```suggestion
   import interpreters

   tasks_recv, tasks = interpreters.create_channel()
```

   results, results_send = interpreters.create_channel()

   def worker():
       interp = interpreters.create()
       interp.set_main_attrs(tasks=tasks_recv, results=results_send)
       interp.exec(tw.dedent("""
           def handle_request(req):
               ...

           def capture_exception(exc):
               ...

           while True:
               try:
                   req = tasks.recv()
               except Exception:
                   # channel closed
                   break
               try:
                   res = handle_request(req)
               except Exception as exc:
                   res = capture_exception(exc)
               results.send_nowait(res)
           """))
   threads = [threading.Thread(target=worker) for _ in range(20)]
   for t in threads:
       t.start()

   requests = ...
   for req in requests:
       tasks.send(req)
   tasks.close()

   for t in threads:
       t.join()

> Sharing a memoryview (imagine map-reduce):

Reviewer 3: ```suggestion
Sharing a ``memoryview`` (imagine map-reduce):
```

                   break
               resindex, start, end = req
               chunk = data[start: end]
               res = reduce_chunk(chunk)
               results[resindex] = res
           """))
   t = threading.Thread(target=worker)
   t.start()

   for i in range(numchunks):
       if not workers_running():
           raise ...
       start = i * chunksize
       end = start + chunksize
       if end > len(buf):
           end = len(buf)
       tasks.send((start, end, i))
   tasks.close()
   t.join()

   use_results(results)

Documentation
=============

The new stdlib docs page for the ``interpreters`` module will include
the following:

* (at the top) a clear note that support for multiple interpreters
  is not required from extension modules
* some explanation about what subinterpreters are
* brief examples of how to use multiple interpreters
  (and communicating between them)
* a summary of the limitations of using multiple interpreters
* (for extension maintainers) a link to the resources for ensuring
  multiple interpreters compatibility
* much of the API information in this PEP

Docs about resources for extension maintainers already exist on the

> `Isolating Extension Modules <isolation-howto_>`_ howto page.  Any
> extra help will be added there.  For example, it may prove helpful
> to discuss strategies for dealing with linked libraries that keep
> their own subinterpreter-incompatible global state.
> 
> .. _isolation-howto:
>    https://docs.python.org/3/howto/isolating-extensions.html

Reviewer 3: We can link using Intersphinx:
```suggestion
:ref:`python:isolating-extensions-howto` howto page.  Any
extra help will be added there.  For example, it may prove helpful
to discuss strategies for dealing with linked libraries that keep
their own subinterpreter-incompatible global state.
```

Also, the ``ImportError`` for incompatible extension modules will be
updated to clearly say it is due to missing multiple interpreters
compatibility and that extensions are not required to provide it.  This
will help set user expectations properly.

Rejected Ideas
==============

> See :pep:`554`.

Reviewer 3: ```suggestion
See :pep:`PEP 554 <554#rejected-ideas>`.
```

state associated with it, though, again, only one may be current.

When there's more than one thread state for a thread,
``PyThreadState_Swap()`` is used to switch between them,
with the requested thread state becoming the current one.
Whatever was running in the thread using the old thread state
is effectively paused until that thread state is swapped back in.

Interpreter States
------------------

As noted earlier, there is some runtime state that multiple threads
share.  Some of it is exposed by the ``sys`` module, though much is
used internally and not exposed explicitly or only through the C-API.

This state that thread states share is called the interpreter state
(``PyInterpreterState``).  We'll sometimes refer to it here as just
"interpreter", though that is also sometimes used to refer to the
``python`` executable, to the Python implementation, and to the
bytecode interpreter (i.e. ``exec()``/``eval()``).

CPython has supported multiple interpreters in the same process (AKA
"subinterpreters") since version 1.5 (1997).  The feature has been
available via the :ref:`C API <python:sub-interpreter-support>`.

> Interpreters and Threads

Author: It relates to how `Interpreter.exec()` and `Interpreter.prepare_main()` affect the current OS thread.  I suppose that is only sort of related to the proposed API.

> Interpreters and Threads

Reviewer: But do they? Sure, `Interpreter.exec()` runs stuff in the current OS thread, but after it's done, everything is the same. I guess if you were looking from another OS thread at which interpreter or `ThreadState` is active in this OS thread (if there even is an API to do so) you'd see the other interpreter.

I don't see how `prepare_main()` affects the current OS thread -- does it? How?

> Interpreters and Threads

Author: My primary concern is that people come in thinking that `Interpreter.exec()` automatically runs in a new thread, which isn't the case.  It's been a persistent confusion.  It's one of the reasons I renamed the method from `run()` to `exec()`.  That's why I've tried to spell out the relationship between interpreters and threads in several places, including this section.  If you aren't seeing the need to clarify relative to `Interpreter.exec()` then I've succeeded. 😄  If you think this section doesn't add much for understanding the execution context then I'd be glad to remove it.

> I don't see how `prepare_main()` affects the current OS thread -- does it? How?

`prepare_main()` must switch interpreters in current OS thread just like `exec()` does.  It has to modify the `__main__` module of the other interpreter, and the shared objects we're binding (at least the ones that aren't directly shareable) must get created in that interpreter.  For both cases we have to switch interpreters in the current OS thread.

> Interpreters and Threads

Reviewer: Oh, I see, in order for `prepare_main()` to even update its interpreter's `__main__` it must make that interpreter the current interpreter. That's worth calling out in the spec for `prepare_main()`, but I don't worry about it the way I (still!) worry about `exec()`.

Regarding `exec()`, I think the problem is that running the code in the current OS thread is just not what people want. Given the purported use cases for subinterpreters, i.e. to utilize other cores, it's just basically never what you want, and the nice short name of the function makes it the first thing people look for.

Maybe it should just be renamed to `exec_in_current_os_thread()`? Or `exec_and_wait()`? Or `exec_synchronously()`, `exec_sync()`, or even `exec_here()`? Even if there will never be a function that spins up a new OS thread where the code is executed (though why not?), the longer name would alert everyone to the semantic pitfall. (And maybe direct them towards the interpreter pool API.)

> Interpreters and Threads

Author: I see what you're saying.  I'll try `exec_sync()`.

I'll also consider adding a `run()` method as the variant that creates a thread to run in.

------------------------

There's a one-to-many relationship between thread states and interpreter
states.  A thread state belongs to a single interpreter and stores
a pointer to it.  That thread state is never used for a different
interpreter.  In the other direction, however, an interpreter may have
more than thread state associated with it.  The interpreter is only
considered active in threads where one of its thread states is current.

Interpreters are created via the C-API using ``Py_NewInterpreter()``
or ``Py_NewInterpreterFromConfig()``.  Both functions do the following:

1. create a new interpreter state
2. create a new thread state
3. set the thread state as current

> 4. initialize the interpreter state using that thread state

Author: That ordering is sort of an implementation detail.  Parts of interpreter init make use of the C-API, which needs a current thread state (e.g. for the current exception, imports).  It was simplest to just say we first set that tstate as current.

I'll add a brief note.

5. return the thread state (still current)

To make an existing interpreter active in the current thread,
we first make sure it has a thread state for the thread.  Then
we call ``PyThreadState_Swap()`` like normal using that thread state.
If the thread state for another interpreter was already current then
it gets swapped out like normal and execution of that interpreter in

> the thread is thus effectively paused until it is swapped back in.

Author: > Now I'm curious what happens if _no_ thread state was already current.

If there wasn't one then nothing's really that different.  The new interpreter get created and its new thread state gets set as current.

> Or if another thread state for the same interpreter was current.

That shouldn't be possible since we're just barely creating the new interpreter.  There wouldn't be any opportunity to create a second thread state for that interpreter.

Once an interpreter is active in the current thread like that, we can
call any of the C-API, such as ``PyEval_EvalCode()`` (i.e. ``exec()``).

The "Main" Interpreter
----------------------

When a Python process starts, it creates a single interpreter state
(the "main" interpreter) with a single thread state for the current
thread.  The Python runtime is then initialized using them.

After initialization, the script or module or REPL is executed using
them.  That execution happens in the interpreter's ``__main__`` module.

When execution is done, the Python runtime is finalized in the main

> thread using the main interpreter.

Author: > I'm guessing that "execution is done" is a way of saying "the main thread of the main interpreter exits", right?

right

> Does it do anything to alert other interpreters and threads?

Right away, finalization waits for all non-daemon threads to finish.  Later it finalizes all subinterpreters, which waits for each of their non-daemon threads to finish.  The various atexit funcs get called pretty early too.

> thread using the main interpreter.

Author: I'll add a note about this.

equivalent (and stay that way).  Thus the proposal does not
include automatic pickling.

Objects vs. ID Proxies
----------------------

For both interpreters and queues, we provide objects that expose them
indirectly with ID-wrapping surrogates for the underlying state.
In both cases the state is process-global and will be used by multiple
interpreters.  Thus they aren't suitable to be implemented as
``PyObject``, which is only really an option for interpreter-specific
data.  That's why the ``interpreters`` module instead provides objects
that are weakly associated through the ID.

Specification
=============

The module will:

* expose the existing multiple interpreter support
* introduce a basic mechanism for communicating between interpreters

The module will wrap a new low-level ``_interpreters`` module
(in the same way as the ``threading`` module).
However, that low-level API is not intended for public use
and thus not part of this proposal.

Using Interpreters
------------------

The module defines the following functions:

* ``get_current() -> Interpreter``
      Returns an ``Interpreter`` object for the currently executing

>       interpreter.

Author: I've pivoted to match `threading`, where there is only one object with a given interpreter ID.  That seems like it would be less surprising.

* ``list_all() -> list[Interpreter]``
      Returns an ``Interpreter`` object for each existing interpreter,
      whether it is currently running in any threads or not.

* ``create() -> Interpreter``
      Create a new interpreter and return an ``Interpreter`` object
      for it.  The interpreter will not be associated with any threads

>       until something is actually run in the interpreter.

Author: Yeah, this is a bit unclear.  My intention was to reinforce the idea that an interpreter on its own isn't associated with any threads.  It isn't until the interpreter is used, e.g. `Interpreter.exec()`, that it gets tied to the current thread.  I'll make that more clear.

Python-Version: 3.13
Replaces: 554

.. note::
   This PEP is essentially a continuation of :pep:`554`.  That document
   had grown a lot of ancillary information across 7 years of discussion.
   This PEP is a reduction back to the essential information.  Much of
   that extra information is still valid and useful, just not in the
   immediate context of the specific proposal here.

Abstract
========

This PEP proposes to add a new module, ``interpreters``, to support
inspecting, creating, and running code in multiple interpreters in the
current process.  This includes ``Interpreter`` objects that represent
the underlying interpreters.  The module will also provide a basic
``Queue`` class for communication between interpreters.  Finally, we
will add a new ``concurrent.futures.InterpreterPoolExecutor`` based
on the ``interpreters`` module.

Introduction
============

Fundamentally, an "interpreter" is the collection of (essentially)
all runtime state which Python threads must share.  So, let's first
look at threads.  Then we'll circle back to interpreters.

Threads and Thread States
-------------------------

A Python process will have one or more threads running Python code

> (or otherwise interacting with the C-API).  Each of these threads

Reviewer 3: Re: https://docs.python.org/3/c-api/index.html

```suggestion
(or otherwise interacting with the C API).  Each of these threads
```

targets a distinct thread state (``PyThreadState``) which holds all
the runtime state unique to that thread.  Each thread also targets
some runtime state it shares with other threads.

A thread's distinct thread state is known as the "current" thread state.
It is stored by the runtime in a thread-local variable, and may be
looked up explicitly with ``PyThreadState_Get()``.  It gets set
automatically for the initial ("main") thread and for

> ``threading.Thread`` objects.  From the C-API it is set (and cleared)
> by ``PyThreadState_Swap()`` and may be set by ``PyGILState_Ensure()``.
> Most of the C-API requires that there be a current thread state,

Reviewer 3: ```suggestion
``threading.Thread`` objects.  From the C API it is set (and cleared)
by ``PyThreadState_Swap()`` and may be set by ``PyGILState_Ensure()``.
Most of the C API requires that there be a current thread state,
```

either looked up implicitly or passed in as an argument.

The relationship between threads and thread states in one-to-many.
Each thread state is associated with at most a single thread and
records its thread ID.  It is never used for another thread.
In the other direction, however, a thread may have more than one thread
state associated with it, though, again, only one may be current.

When there's more than one thread state for a thread,
``PyThreadState_Swap()`` is used to switch between them,
with the requested thread state becoming the current one.
Whatever was running in the thread using the old thread state
is effectively paused until that thread state is swapped back in.

Interpreter States
------------------

As noted earlier, there is some runtime state that multiple threads
share.  Some of it is exposed by the ``sys`` module, though much is

> used internally and not exposed explicitly or only through the C-API.

Reviewer 3: ```suggestion
used internally and not exposed explicitly or only through the C API.
```

(the "main" interpreter) with a single thread state for the current
thread.  The Python runtime is then initialized using them.

After initialization, the script or module or REPL is executed using
them.  That execution happens in the interpreter's ``__main__`` module.

When execution is done, the Python runtime is finalized in the main
thread using the main interpreter.

Interpreter Isolation
---------------------

CPython's interpreters are intended to be strictly isolated from each
other.  That means interpreters never share objects (except in very
specific cases internally). Each interpreter has its own modules
(``sys.modules``), classes, functions, and variables.  Even where
two interpreters define the same class, each will have a distinct copy.
The same applies to state in C, including in extension modules.
The CPython C-API docs `explain more`_.

.. _explain more:
   https://docs.python.org/3/c-api/init.html#bugs-and-caveats

Notably, there is some process-global state that interpreters will
always share:

* file descriptors
* low-level env vars
* process memory (though allocators *are* isolated)
* builtin types (e.g. dict, bytes)

> * singletons (e.g. None)

Reviewer 3: ```suggestion
* singletons (e.g. ``None``)
```

between interpreters may be sent through a ``Queue``.  Note that the
actual objects aren't sent, but rather their underlying data.
However, the popped object will still be identical to the original.
See `Shareable Objects`_.

The module defines the following functions:

* ``create_queue(maxsize=0) -> Queue``
      Create a new queue.  If the maxsize is zero or negative then the
      queue is unbounded.

Queue Objects
-------------

``interpreters.Queue`` objects act as proxies for the underlying
cross-interpreter-safe queues exposed by the ``interpreters`` module.

``Queue`` implements all the methods of ``queue.Queue`` except for
``task_done()`` and ``join()``, hence it is similar to
``asyncio.Queue`` and ``multiprocessing.Queue``.

Attributes and methods:

* ``id``
      (read-only) The target queue's ``QueueID`` object.
      It is an int-like object with some internal bookkeeping.

* ``maxsize``
      Number of items allowed in the queue.

* ``__hash__()``
      Returns the hash of the queue's ``id``.  This is the same
      as the hash of the ID's integer value.

* ``__eq__(other)``
      Returns ``other is self``.

* ``empty()``

>       Return True if the queue is empty, False otherwise.

Reviewer 3: ```suggestion
      Return ``True`` if the queue is empty, False otherwise.
```

* ``full()``

>       Return True if there are maxsize items in the queue.
> 
>       If the queue was initialized with maxsize=0 (the default),
>       then full() never returns True.

Reviewer 3: ```suggestion
      Return ``True`` if there are ``maxsize`` items in the queue.

      If the queue was initialized with ``maxsize=0`` (the default),
      then ``full()`` never returns ``True``.
```

* ``qsize()``
      Return the number of items in the queue.

* ``put(obj, timeout=None)``
      Add the object to the queue.

      The object must be `shareable <Shareable Objects_>`_, which means
      the object's data is passed through rather than the object itself.

>       If maxsize > 0 and the queue is full then this blocks until a free
>       slot is available.  If "timeout" is a positive number then it only

Reviewer 3: Italics for params: https://devguide.python.org/documentation/markup/#quick-reference

```suggestion
      If ``maxsize > 0`` and the queue is full then this blocks until a free
      slot is available.  If *timeout* is a positive number then it only
```

      blocks at least that many seconds and then raises
      ``interpreters.QueueFull``.  Otherwise is blocks forever.

* ``put_nowait(obj)``
      Like ``put()`` but effectively with an immediate timeout.

* ``get(timeout=None) -> object``
      Pop the next object from the queue and return it.  Block while

>       the queue is empty.  If a positive timeout is provided and an

Reviewer 3: ```suggestion
      the queue is empty.  If a positive *timeout* is provided and an
```

corresponding objects in different interpreters will always strictly
match each other.

Synchronization
---------------

There are situations where two interpreters should be synchronized.
That may involve sharing a resource, worker management, or preserving
sequential consistency.

In threaded programming the typical synchronization primitives are
types like mutexes.  The ``threading`` module exposes several.
However, interpreters cannot share objects which means they cannot
share ``threading.Lock`` objects.

The ``interpreters`` module does not provide any such dedicated
synchronization primitives.  Instead, ``Queue`` objects provide
everything we need.

For example, if there's a shared resource that needs managed
access then a queue may be used to manage it, where the interpreters
pass an object around to indicate who can use the resource::

   import interpreters
   from mymodule import load_big_data, check_data

   numworkers = 10
   control = interpreters.create_queue()
   data = memoryview(load_big_data())

   def worker():
       interp = interpreters.create()
       interp.prepare___main__(control=control, data=data)

>        interp.exec("""if True:

Reviewer 3: Looks like we're missing the end of the triple-quoted strings in these two examples?

It's more obvious from the preview:

https://pep-previews--3523.org.readthedocs.build/pep-0734/#synchronization

(in the same way as the ``threading`` module).
However, that low-level API is not intended for public use
and thus not part of this proposal.

Using Interpreters
------------------

The module defines the following functions:

* ``get_current() -> Interpreter``
      Returns an ``Interpreter`` object for the currently executing
      interpreter.

* ``list_all() -> list[Interpreter]``
      Returns an ``Interpreter`` object for each existing interpreter,
      whether it is currently running in any threads or not.

* ``create() -> Interpreter``
      Create a new interpreter and return an ``Interpreter`` object
      for it.  The interpreter will not be associated with any threads
      until something is actually run in the interpreter.

Interpreter Objects
-------------------

An ``interpreters.Interpreter`` object represents the interpreter
(``PyInterpreterState``) with the corresponding ID.

If the interpreter was created with ``interpreters.create()`` then
it will be destroyed as soon as all ``Interpreter`` objects have been
deleted.

Attributes and methods:

* ``id``
      (read-only) The target interpreter's ``InterpreterID`` object.

>       It is an int-like object with some internal bookkeeping.

Reviewer: Why does it have to be a special object? What does the extra bookkeeping do?

Author: Yeah, this doesn't need to be a special object.  I'll change it to `int`.

      between interpreters are allowed.  See `Shareable Objects`_.

      ``prepare___main__()`` is helpful for initializing the
      globals for an interpreter before running code in it.

* ``exec(code, /)``
      Run the given source code in the interpreter
      (in the current thread).

      This is essentially equivalent to switching to this interpreter
      in the current thread and then calling the builtin ``exec()``
      using this interpreter's ``__main__`` module's ``__dict__`` as
      the globals and locals.

      ``exec()`` does not reset the interpreter's state nor
      the ``__main__`` module, neither before nor after, so each
      successive call picks up where the last one left off.  This can
      be useful for running some code to initialize an interpreter
      (e.g. with imports) before later performing some repeated task.

Comparison with builtins.exec()
-------------------------------

``Interpreter.exec()`` is essentially the same as the builtin
``exec()``, except it targets a different interpreter, using that
interpreter's distinct runtime state.

Here are the relevant characteristics of the builtin ``exec()``,
for comparison:

> * It runs in the current OS thread and pauses whatever was running there,

Reviewer: I keep getting confused about whether "it" refers to builtin.exec() or to Interpreter.exec() or both. I recommend being explicit and saying "Both ..." if that's what you mean (though some of the bullets seem to be calling out differences.)

  which resumes when ``exec()`` finishes.  No other threads are affected.
  (To avoid pausing the current thread, run ``exec()``
  in a ``threading.Thread``.)
* It executes against a namespace, by default the ``__dict__`` of the

>   current module (i.e. ``globals()``).  It uses that namespace as-is

Reviewer: The default for builtins.exec() is not always globals() -- when run inside a function it is locals(). I think Interpreter.exec() always runs against the namespace of *that interpreter's* `__main__` module, right?

Author: Good catch!

  and does not clear it before or after.
* When code is run from the command-line (e.g. ``-m``) or the REPL,
  the "current" module is always the ``__main__`` module

>   of the target (main) interpreter.

Reviewer: How is this relevant to Interpreter.exec()?

Reviewer: That feels like a stretch -- I wouldn't say that `exec()` is like `python -c` either. (E.g., `-c` prints the traceback, but `exec()` doesn't -- the traceback printing is in a different part of the functionality.)

Rather than trying to use analogies, I think it would work better if you just specified the intended behavior carefully.

Author: > That feels like a stretch -- I wouldn't say that `exec()` is like `python -c` either. (E.g., `-c` prints the traceback, but `exec()` doesn't -- the traceback printing is in a different part of the functionality.)

FWIW, I am moving this up to the Introduction and converting this specific bullet into a distinct, brief discussion of how the different CLI options relate to the relevant C API functions (e.g. `PyRun_String()`, `PyEval_EvalCode()`).  That's in the interest of identifying how `exec()`, `Interpreter.exec()`, and the CLI are related at a fundamental level.

> Rather than trying to use analogies, I think it would work better if you just specified the intended behavior carefully.

Fair enough.  I'll clean that up.

> * It discards any object returned from the executed code::

Reviewer: Does this mean you can actually pass `func.__code__` to Interpreter.exec()? How does that work? If that's not supported, I'm confused -- why bring this up?

Author: I plan on supporting that right away, but didn't want to clutter the PEP with another dimension of usage.  I can remove this bullet if you think it's a distraction.

Reviewer: Have you implemented that yet? How were you planning to pass the code object around? Using marshal perhaps?

Author: Yeah, the implementation was surprisingly trivial.  I did indeed use marshal.

(FYI, I'm removing the bullet.)

   def func():
       global spam
       spam = True
       return 'a returned value'

   ns = {}
   res = exec(func.__code__, ns)
   # Nothing is returned.  The returned string was discarded.
   assert res is None, res
   assert ns['spam'] is True, ns

* It propagates any uncaught exception from the code it ran.
  The exception is raised from the ``exec()`` call in the thread
  that originally called ``exec()``.

> The only difference is that, rather than propagate the uncaught

Reviewer: This paragraph seems logically to be part of the last bullet. Why not indent it accordingly?

Author: I'm moving the paragraph elsewhere.

(including its traceback) as the ``__cause__``.  That means if the
``RunFailedError`` isn't caught then the full traceback of the
propagated exception, including details about syntax errors, etc.,
will be displayed.  Having the full traceback is particularly useful
when debugging.

Communicating Between Interpreters
----------------------------------

The module introduces a basic communication mechanism through special
queues.

There are ``interpreters.Queue`` objects, but they only proxy
the actual data structure: an unbounded FIFO queue that exists
outside any one interpreter.  These queues have special accommodations
for safely passing object data between interpreters, without violating
interpreter isolation.  This includes thread-safety.

As with other queues in Python, for each "put" the object is added to
the back and each "get" pops the next one off the front.  Every added
object will be popped off in the order it was pushed on.

Only objects that are specifically supported for passing
between interpreters may be sent through a ``Queue``.  Note that the
actual objects aren't sent, but rather their underlying data.

> However, the popped object will still be identical to the original.

Reviewer: "Identical" is probably not the term you're looking for here -- that would imply the two have the same address. I'd use equivalent or equal instead.

See `Shareable Objects`_.

The module defines the following functions:

* ``create_queue(maxsize=0) -> Queue``
      Create a new queue.  If the maxsize is zero or negative then the
      queue is unbounded.

Queue Objects
-------------

``interpreters.Queue`` objects act as proxies for the underlying
cross-interpreter-safe queues exposed by the ``interpreters`` module.

``Queue`` implements all the methods of ``queue.Queue`` except for
``task_done()`` and ``join()``, hence it is similar to
``asyncio.Queue`` and ``multiprocessing.Queue``.

Attributes and methods:

* ``id``
      (read-only) The target queue's ``QueueID`` object.
      It is an int-like object with some internal bookkeeping.

* ``maxsize``

>       Number of items allowed in the queue.

Reviewer: I'd add "Zero means unbounded."

* ``__hash__()``
      Return the hash of the queue's ``id``.  This is the same
      as the hash of the ID's integer value.

* ``__eq__(other)``

>       Return ``other is self``.

Reviewer: That's just inherited from `object`? Then it's not worth listing here. But maybe it should compare the `id` field?

Author: Good point.  It was previously comparing `.id` but I changed it to identity later.  I'll drop it.

* ``empty()``
      Return ``True`` if the queue is empty, ``False`` otherwise.

* ``full()``
      Return ``True`` if there are ``maxsize`` items in the queue.

      If the queue was initialized with ``maxsize=0`` (the default),
      then ``full()`` never returns ``True``.

* ``qsize()``

>       Return the number of items in the queue.

Reviewer: Maybe mention that these three are snapshots and their value can change any moment? E.g. qsize() may return 1 and yet get() could block -- if someone else snatched the item out of the queue before you.

* ``put(obj, timeout=None)``
      Add the object to the queue.

      The object must be `shareable <Shareable Objects_>`_, which means
      the object's data is passed through rather than the object itself.

      If ``maxsize > 0`` and the queue is full then this blocks until
      a free slot is available.  If *timeout* is a positive number
      then it only blocks at least that many seconds and then raises
      ``interpreters.QueueFull``.  Otherwise is blocks forever.

* ``put_nowait(obj)``

>       Like ``put()`` but effectively with an immediate timeout.

Reviewer: That differs from queue.Queue, which raises Full if the queue is full.

Author: That's what I was trying to communicate.  I'll clarify the text.

* ``get(timeout=None) -> object``
      Pop the next object from the queue and return it.  Block while
      the queue is empty.  If a positive *timeout* is provided and an
      object hasn't been added to the queue in that many seconds
      then raise ``interpreters.QueueEmpty``.

> * ``get_nowait([default]) -> object``

Reviewer: Any particular reason to add the default arg?

Author: It's a small thing but I personally found it useful.  If you think it's a distraction I could drop it.

Reviewer: I'd like as much API knowledge as possible to transfer between the different queue types we have (asyncio, threading, subinterpreters).

Author: You mean "keep the APIs as similar as possible"?  That's reasonable.

      Like ``get()``, but do not block.  If the queue is not empty
      then return the next item.  Otherwise, return *default* if
      one was provided else raise ``interpreters.QueueEmpty``.

Shareable Objects
-----------------

Both ``Interpreter.prepare___main__()`` and ``Queue`` work only with
"shareable" objects.

A "shareable" object is one which may be passed from one interpreter
to another.  The object is not necessarily actually shared by the
interpreters.  However, the object in the one interpreter is guaranteed
to exactly match the corresponding object in the other interpreter.

For some types (builtin singletons), the actual object is shared.
For some, the object's underlying data is actually shared but each

> interpreter has a distinct object wrapping that data.  For all other

Reviewer: I guess that "wrapping" only happens for memoryview?

Author: yep

corresponding objects continue to match exactly.  In cases where
the underlying data is complex but must be copied (e.g. ``tuple``),
the data is serialized as efficiently as possible.

Shareable objects must be specifically supported internally
by the Python runtime.  However, there is no restriction against
adding support for more types later.

Here's the initial list of supported objects:

* ``str``
* ``bytes``
* ``int``
* ``float``
* ``bool`` (``True``/``False``)
* ``None``
* ``tuple`` (only with shareable items)
* ``Queue``
* ``memoryview`` (underlying buffer actually shared)

Note that the last two on the list, queues and ``memoryview``, are
technically mutable data types, whereas the rest are not.  When any
interpreters share mutable data there is always a risk of data races.
Cross-interpreter safety, including thread-safety, is a fundamental
feature of queues.

However, ``memoryview`` does not have any native accommodations.
The user is responsible for managing thread safety, whether passing
a token back and forth through a queue to indicate safety
(see `Synchronization`_), or by assigning sub-range exclusivity
to individual interpreters.

> Regarding queues, the primary situation for sharing is where an

Reviewer: Maybe add "(described below)"?

interpreter is using ``prepare___main__()`` to provide the target
interpreter with their means of further communication.

Finally, a reminder: for some types the actual object is shared,
whereas for others only the underlying data (or even a copy or proxy)
is shared.  Regardless, the guarantee of "shareable" objects is that
corresponding objects in different interpreters will always strictly

> match each other.

Reviewer: I guess these are weasel words because you can't use `__eq__` and for NaNs that wouldn't even work. :-) I assume unicode normalization isn't going to be a problem.

Author: > I guess these are weasel words because you can't use `__eq__` and for NaNs that wouldn't even work. :-)

By "strictly match" I meant a deeper sort of equivalence than just equality.  The "shared" object would be indistinguishable from the original, whether it were the actual original object or a copy/proxy.  For oddballs like `NaN` or `inf` you'd get the right object on the other side (i.e.. `math.isnam()` and `math.isinf()`, respectively, would return true).  For mutable data, like buffers, mutation in one interpreter would be reflected in every other interpreter in which the object had been shared through a queue (or `prepare_main()`).

I'm definitely open to clearer phrasing.  In the meantime, I'll change it to "strictly equivalent".  Regardless, I'll try clarifying things with some additional text.

> I assume unicode normalization isn't going to be a problem.

Yeah, it shouldn't be a problem.

Reviewer: I think being explicit about what you just told me in the comment would be good. Especially the part about mutable buffers being (or acting as) truly shared. (Though all the C code manipulating these probably will need to use atomic operations to ensure the compiler actually writes to the buffer when you expect it to.)

       interp.prepare___main__(sync=sync)
       interp.exec("""if True:
           # do step 1
           token = sync.get()
           sync.put(token)
           # do step 2
           sync.put(None)
           sync.get()
           # do step 3
           sync.put(None)
           sync.get()
           """)
   t = threading.Thread(target=worker)
   t.start()

   # do step A
   sync.put(None)
   sync.get()
   # do step B
   token = sync.get()
   sync.put(token)
   # do step C
   token = sync.get()
   sync.put(token)

Exceptions
----------

* ``RunFailedError``
      Raised from ``Interpreter.exec()`` when there's an uncaught
      exception.

      This exception is a subclass of ``RuntimeError``.

* ``QueueEmpty``
      Raised from ``Queue.get()`` (or ``get_nowait()`` with no default)
      when the queue is empty.

      This exception is a subclass of ``queue.Empty``.

* ``QueueFull``
      Raised from ``Queue.put()`` (with a timeout) or ``put_nowait()``
      when the queue is already at its max size.

      This exception is a subclass of ``queue.Full``.

InterpreterPoolExecutor
-----------------------

The new ``concurrent.futures.InterpreterPoolExecutor`` will be

> a subclass of ``concurrent.futures.ThreadPoolExecutor``, where each

Reviewer: Maybe just promise `c.f.Executor`? That's an abstract type that defines the interface.

   threads = []

   def results_handler():
       running = numworkers
       while running:
           try:
               res = results.get(timeout=0.1)
           except interpreters.QueueEmpty:
               # No workers have finished a request since last time.
               pass
           else:
               if res is None:
                   # A worker has stopped.
                   running -= 1
               else:
                   handle_result(res)
       empty = object()
       assert results.get_nowait(empty) is empty
   threads.append(threading.Thread(target=results_handler))

   def worker():
       interp = interpreters.create()
       interp.prepare___main__(tasks=tasks, results=results)
       interp.exec("""if True:
           from mymodule import handle_request, capture_exception

           while True:
               req = tasks.get()
               if req is None:
                   # Stop!
                   break
               try:
                   res = handle_request(req)
               except Exception as exc:
                   res = capture_exception(exc)
               results.put_nowait(res)
           # Notify the results handler.

>            results.put_nowait(None)

Reviewer: Again, I'd just use `put()` everywhere.

>            """))

Reviewer: Extra ')'? Also in next example.

There are situations where two interpreters should be synchronized.
That may involve sharing a resource, worker management, or preserving
sequential consistency.

In threaded programming the typical synchronization primitives are
types like mutexes.  The ``threading`` module exposes several.
However, interpreters cannot share objects which means they cannot
share ``threading.Lock`` objects.

The ``interpreters`` module does not provide any such dedicated
synchronization primitives.  Instead, ``Queue`` objects provide
everything we need.

For example, if there's a shared resource that needs managed
access then a queue may be used to manage it, where the interpreters
pass an object around to indicate who can use the resource::

   import interpreters
   from mymodule import load_big_data, check_data

   numworkers = 10
   control = interpreters.create_queue()
   data = memoryview(load_big_data())

   def worker():
       interp = interpreters.create()
       interp.prepare___main__(control=control, data=data)
       interp.exec("""if True:
           from mymodule import edit_data
           while True:
               token = control.get()
               edit_data(data)

>                control.put_nowait(token)

Author: good point
